libs/libcds/include/hpx/libcds/hpx_tls_manager.hpp

```diff
@@ -0,0 +1,249 @@
+
+        // the boolean uselibcs option is provided to make comparison
+        // of certain tests with/without libcds easier
+        // @TODO : we should remove it one day
+        explicit hpxthread_manager_wrapper(bool uselibcds = true)
+          : uselibcds_(uselibcds)
+        {
+            if (uselibcds_)
+            {
+                if (++thread_counter_ >
+                    std::stoul(hpx::get_config_entry(
+                        "hpx.cds.num_concurrent_hazard_pointer_threads",
+                        "128")))
+                {
+                    HPX_THROW_EXCEPTION(invalid_status,
+                        "hpx::cds::hpxthread_manager_wrapper ",
+                        "attaching more threads than number of maximum allowed "
+                        "detached threads, consider update "
+                        "--hpx:ini=hpx.cds.num_concurrent_hazard_pointer_"
+                        "threads to a larger number");
+                }
+
+                if (::cds::gc::hp::custom_smr<
+                        ::cds::gc::hp::details::HPXDataHolder>::isUsed())
+                {
+                    ::cds::gc::hp::custom_smr<
+                        ::cds::gc::hp::details::HPXDataHolder>::attach_thread();
+                }
+                else
+                {
+                    HPX_THROW_EXCEPTION(invalid_status,
+                        "hpx::cds::hpxthread_manager_wrapper ",
+                        "failed to attach_thread to HPXDataHolder, please check"
+                        "if hazard pointer is constructed.");
+                }
+            }
+        }
+
+        ~hpxthread_manager_wrapper()
+        {
+            if (uselibcds_)
+            {
+                if (thread_counter_-- == 0)
+                {
+                    HPX_THROW_EXCEPTION(invalid_status,
+                        "hpx::cds::hpxthread_manager_wrapper",
+                        "detaching more threads than number of attached "
+                        "threads");
+                }
+                ::cds::gc::hp::custom_smr<
+                    ::cds::gc::hp::details::HPXDataHolder>::detach_thread();
+            }
+        }
+
+        // max_concurrent_attach_thread is corresponding variable to
+        // the variable nMaxThreadCount in Hazard Pointer class. It defines
+        // max count of simultaneous working thread in the application, default 100
+        // and it is public to user for use
+
+    private:
+        static std::atomic<std::size_t> thread_counter_;
+        bool uselibcds_;
+    };
+
+    // this wrapper will initialize a std::thread for use with libCDS
+    // algorithms
+    struct stdthread_manager_wrapper
+    {
+        explicit stdthread_manager_wrapper()
+        {
+            if (++hpxthread_manager_wrapper::thread_counter_ >
+                std::stoul(hpx::get_config_entry(
+                    "hpx.cds.num_concurrent_hazard_pointer_threads", "128")))
```

Comment:
This is not the end of the world, but I would prefer to see this be independent of the runtime configuration. If this really needs to have a default value set like this we can install a callback to get the configuration value instead. Otherwise what is done here: https://github.com/STEllAR-GROUP/hpx/blob/125de3907b128c4ae3972642a9b9796253961a9b/libs/libcds/tests/unit/libcds_map_hazard_pointer_hpxthread.cpp#L163-L167 is a fine alternative, but it does push the responsibility to the user.

Reply:
I can hardcode n_threads = 100 (it was true several commits old), would you prefer this way?